---
title: "RNAed_notebook"
author: "ab"
date: "November 27, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r Get data, echo = FALSE}
#Attach libraries
require(tidyverse)

#Define functions
numextract.int <- function(string){ 
  str_extract(string, "\\-*\\d*\\d*")
} 
numextract.double <- function(string){ 
  str_extract(string, "\\-*\\d+\\.*\\d*")
} 
Load.Data <- function(Path, Save){
  #Establish filepaths
  Path.Data <- Path
  Files <- list.files(file.path(Path.Data), pattern = "*.vcf$")
  Meta <- read_csv("FANTOM6_samples.csv", col_names = TRUE)
  #Merging loop
  for (i in 1:length(Files)) {
    #Read vcf
    tmp.vcf <- readLines(file.path(Path.Data, Files[i]))
    tmp.vcf.data <- read.table(file.path(Path.Data, Files[i]))
    #Extract and atach colnames
    tmp.vcf <- tmp.vcf[-(grep("#CHROM", tmp.vcf) + 1):-(length(tmp.vcf))]
    vcf.names <- unlist(strsplit(tmp.vcf[length(tmp.vcf)], "\t"))
    names(tmp.vcf.data)<-vcf.names
    #Extract DP, EF and Metadata
    tmp.vcf.data <- tmp.vcf.data %>% 
      mutate(DP = INFO %>% 
               str_split_fixed(., pattern = ";", n = 2) %>% 
               as.tibble() %>%
               pull(V1) %>%
               numextract.double(.) %>%
               as.integer(), 
             EF = INFO %>% 
               str_split_fixed(., pattern = ";", n = 2) %>% 
               as.tibble() %>%
               pull(V2) %>%
               numextract.double(.) %>%
               as.numeric(),
             EDIT = paste(REF, ALT, sep = " to "),
             File.Name = rep(paste(Files[i]), length(.$`#CHROM`))) %>%
      mutate(SampleID = File.Name %>% 
               str_sub(., 1, -5), 
             EDIT = paste(REF, ALT, sep = " to ")
      ) %>%
      select(-INFO, -FORMAT, -File.Name, -starts_with("RDhi", ignore.case = FALSE))
    if (i == 1) {
      Data <- tmp.vcf.data
      rm(tmp.vcf.data)
    } else {
      if (i > 1 & i < length(Files)) {
        Data <- rbind(Data, tmp.vcf.data) 
      } else {
        if (i == length(Files)) {
          Data <- rbind(Data, tmp.vcf.data)
          #Create Factorial EF variable
          Data <- Data %>% 
            mutate(EF.Factor = EF)
          Data[Data$EF <= 0.25, ]$EF.Factor <- rep("4.LowEf", length(Data[Data$EF <= 0.25, ]$EF.Factor))
          Data[Data$EF > 0.25 & Data$EF <= 0.5, ]$EF.Factor <- rep("3.MidEf Left", length(Data[Data$EF > 0.25 & Data$EF <= 0.5, ]$EF.Factor))
          Data[Data$EF > 0.5 & Data$EF <= 0.75, ]$EF.Factor <- rep("2.MidEf Right", length(Data[Data$EF > 0.5 & Data$EF <= 0.75, ]$EF.Factor))
          Data[Data$EF > 0.75, ]$EF.Factor <- rep("1.HighEf", length(Data[Data$EF > 0.75, ]$EF.Factor))
          #Attach metadata
          Data <- Data %>%
            inner_join(., Meta, by = c("SampleID" = "Sample ID"))
          #Substitute NA for 'Control'
          Data$`KD gene ID`[is.na(Data$`KD gene ID`)] <- "Control"
          Data <- Data %>%
            as.tibble()
          return(Data)
          if(Save == TRUE){
            write_tsv(Data, file.path(Path.Data, "vcf_comp.tsv"), col_names = TRUE)
          }
          rm(tmp.vcf.data, vcf.names, tmp.vcf, i)
        }
      }
    }
  }
}
#Chromosome-wise plotting
Plot.EF.Relative <- function(Cell_Type, Depth_Cut, EdEvents_Cut) {
  
  Cell.Type <- c("ADSC", "BEC", "CN", "HDF (neonatal)", "FF-IPSC (IMS)", "iPSC (YJK)", "KER", "LEC", "NPC (YJK)")
  
  Cell.Def <- as.tibble(table(Data$`Cell description`)) %>% 
    pull(Var1)
  
  temp <- Data %>% 
  filter(`Cell type ID` %in% Cell.Type[Cell_Type] & DP > Depth_Cut) %>%
  group_by(EF.Factor, `#CHROM`, `KD gene ID`, EDIT) %>%
  summarise(counts = n()) %>%
  filter(EF.Factor %in% c("1.HighEf", "4.LowEf") & counts > EdEvents_Cut)
  
  
Data.Rel <- tibble()
for (z in 1:length(table(temp$EDIT))) {
  temp.1 <- as.tibble(table(temp$EDIT))
  temp.2 <- temp %>%
    filter(EDIT == temp.1$Var1[z]) 
  for (x in 1:length(table(temp$`KD gene ID`))) {
  temp.3 <- as.tibble(table(temp$`KD gene ID`))
  temp.4 <- temp.2 %>%
    filter(`KD gene ID` == temp.3$Var1[x]) 
  for (y in 1:length(table(temp$`#CHROM`))) {
  temp.5 <- as.tibble(table(temp$`#CHROM`))
  temp.6 <- temp.4 %>%
    filter(`#CHROM` == temp.5$Var1[y])
  value <- round(as.numeric(temp.6[1, 5] / (temp.6[2, 5] + temp.6[1, 5])), 3)
  value.count.ef <- as.numeric(temp.6[1, 5])
  value.count.tot <- as.numeric(temp.6[1, 5] + temp.6[2, 5])
  temp.7 <- tibble(`#CHROM` = temp.5$Var1[y], 
                   `KD gene ID` = temp.3$Var1[x],
                   EDIT = temp.1$Var1[z],
                   EF.Relative = value, 
                   Event.Count.Tot = value.count.tot,
                   Event.Count.Ef = value.count.ef)
  Data.Rel <- rbind(Data.Rel, temp.7)
  }
  if (x == length(table(temp$`KD gene ID`))) {
    rm(temp.1, temp.2, temp.3, temp.4, temp.5, temp.6, temp.7, x, y, z)
    }
  } 
}

return(
  ggplot(data = Data.Rel) + 
    geom_col(aes(x = `#CHROM`, y = EF.Relative, fill = Event.Count.Tot), 
           position = "dodge") + 
    scale_fill_gradient2(high = "red", low = "blue", mid = "yellow", midpoint = 1000) +
    facet_grid(`KD gene ID` ~ EDIT) +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12)) +
    labs(title = paste(Cell.Def[Cell_Type]))
)
}
Plot.EF.Absolute <- function(Cell_Type, Depth_Cut) {
  
  Cell.Type <- as.tibble(table(Data$`Cell type ID`)) %>%
    pull(Var1)
  
  Cell.Def <- as.tibble(table(Data$`Cell description`)) %>% 
    pull(Var1)
  
  return(
    Data %>% 
      filter(`Cell type ID` %in% Cell.Type[Cell_Type] & DP > Depth_Cut) %>% 
      ggplot(data = .) + 
      geom_bar(aes(x = `#CHROM`, y = ..count.., fill = EF.Factor)) + 
      scale_fill_gradientn(colours = terrain.colors(50))+ 
      facet_grid(`KD gene ID` ~ EDIT) + 
      theme_bw() + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12)) + 
      labs(title = paste(Cell.Def[Cell_Type]))
  )
}
Plot.EF.Dist <- function(Cell_Type, Depth_Cut) {
  
  Cell.Type <- c("ADSC", "BEC", "CN", "HDF (neonatal)", "FF-IPSC (IMS)", "iPSC (YJK)", "KER", "LEC", "NPC (YJK)")
  
  Cell.Def <- as.tibble(table(Data$`Cell description`)) %>% 
    pull(Var1)
  
  Plot <- ggplot(data = Data %>% 
                   filter(`Cell type ID` == Cell.Type[Cell_Type] & DP >= Depth_Cut)) + 
    geom_density(aes(x = EF, fill = `KD gene ID`), alpha = 1/2) +
    facet_grid(EDIT ~ .) + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12)) + 
    labs(title = paste(Cell.Def[Cell_Type]))
  return(Plot)
  
  facet_grid(SM.Type ~ .) +
  theme_bw()
  
  
}
```

```{r explore&plot, echo = FALSE}
Data <- Load.Data("Data/", TRUE)

```


```{r explore further, echo = FALSE}
#Plotting
#For function calls
# Arg.value:  Cell Type:      KD present:
# 1           ADSC            FALSE
# 2           BEC             FALSE
# 3           CN              FALSE
# 4           iPSC (YJK)      FALSE
# 5           KER             FALSE
# 6           LEC             FALSE
# 7           NPC (EJK)       FALSE
# 8           FF-iPSC (IMS)   TRUE
# 9           HDF (neonatal)  TRUE

#EF.Factorial (Factorial Editing Fraction) is defined as:
#1.HighEf = assigned to samples with editing fraction > 0.75
#2.MidEf Right = sampels with editing fraction 0.25 < EF <= 0.5
#3.MedEf Left = sampels with editing fractino 0.5 < EF <= 0.75
#4.LowEf = samples with editing fraction <= 0.25 

#EF.Relative (Relative Editing Fraction) is defined as:
#EF.Relative = count(HighEf events) / count(LowEf events)


#Check density
Plot.EF.Dist(5, 20)

```


```{r more plotting, echo=FALSE}

Plot.EF.Relative(Cell_Type = 4, Depth_Cut = 20, EdEvents_Cut = 5)

```

```{r functional}
Data %>%
  filter(DP > 30 & EF.Factor == "1.HighEf") %>%
  mutate(POS1 = POS) %>%
  select(`#CHROM`, POS, POS1, `KD gene ID`, `Cell type ID`, SampleID) %>%
  write_tsv(., "bed/Data.bed", col_names = TRUE)
#Intersect beds.....
#Read intersections
Data.Genes <- read_tsv("bed/Data.WholeGenes.txt", col_names = FALSE) %>%
  select(X1, X2, X9, X4, X5) %>%
  `colnames<-`(c("`#CHROM`", "POS", "ENSTID", "KD_gene_ID", "Cell_type_ID")) %>%
  mutate(ENSTID = str_replace(ENSTID, 
                              pattern = ".[0-9]+$",
                              replacement = ""
                              )
         )

#Change EnsID to hgncID
require(biomaRt)
Transcripts <- Data.Genes %>%
  dplyr::select(ENSTID) %>%
  distinct()
Mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
Gene.List <- getBM(filters = "ensembl_transcript_id", 
                   attributes = c("ensembl_transcript_id","hgnc_symbol"),
                   values = Transcripts$ENSTID, 
                   mart = Mart
)
rm(Mart, Transcripts)
detach("package:biomaRt", unload=TRUE)


Data.Genes <- Data.Genes %>% 
  left_join(., Gene.List, by = c("ENSTID" = "ensembl_transcript_id"))

Temp <- Data.Genes %>% 
  group_by(Cell_type_ID, KD_gene_ID, hgnc_symbol) %>%
  summarise(Counts = n()) %>%
  spread(key = hgnc_symbol, value = Counts)
  


```

