---
title: "RNAed_Report"
author: "ab"
date: "January 16, 2019"
output:
  html_document:
    df_print: paged
---
RNAediting landscape in FANTOM6 transcriptome profiling


```{r Get Data, echo = FALSE}
#Attach libraries
require(tidyverse)
#Define functions
numextract.int <- function(string){ 
  str_extract(string, "\\-*\\d*\\d*")
} 
numextract.double <- function(string){ 
  str_extract(string, "\\-*\\d+\\.*\\d*")
} 
Load.Data <- function(Path, Save){
  #Establish filepaths
  Path.Data <- Path
  Files <- list.files(file.path(Path.Data), pattern = "*.vcf$")
  Meta <- read_csv("FANTOM6_samples.csv", col_names = TRUE)
  #Merging loop
  for (i in 1:length(Files)) {
    #Read vcf
    tmp.vcf <- readLines(file.path(Path.Data, Files[i]))
    tmp.vcf.data <- read.table(file.path(Path.Data, Files[i]))
    #Extract and atach colnames
    tmp.vcf <- tmp.vcf[-(grep("#CHROM", tmp.vcf) + 1):-(length(tmp.vcf))]
    vcf.names <- unlist(strsplit(tmp.vcf[length(tmp.vcf)], "\t"))
    names(tmp.vcf.data)<-vcf.names
    #Extract DP, EF and Metadata
    tmp.vcf.data <- tmp.vcf.data %>% 
      mutate(DP = INFO %>% 
               str_split_fixed(., pattern = ";", n = 2) %>% 
               as.tibble() %>%
               pull(V1) %>%
               numextract.double(.) %>%
               as.integer(), 
             EF = INFO %>% 
               str_split_fixed(., pattern = ";", n = 2) %>% 
               as.tibble() %>%
               pull(V2) %>%
               numextract.double(.) %>%
               as.numeric(),
             EDIT = paste(REF, ALT, sep = " to "),
             File.Name = rep(paste(Files[i]), length(.$`#CHROM`))) %>%
      mutate(SampleID = File.Name %>% 
               str_sub(., 1, -5), 
             EDIT = paste(REF, ALT, sep = " to ")
      ) %>%
      select(-INFO, -FORMAT, -File.Name, -starts_with("RDhi", ignore.case = FALSE))
    if (i == 1) {
      Data <- tmp.vcf.data
      rm(tmp.vcf.data)
    } else {
      if (i > 1 & i < length(Files)) {
        Data <- rbind(Data, tmp.vcf.data) 
      } else {
        if (i == length(Files)) {
          Data <- rbind(Data, tmp.vcf.data)
          #Create Count variable
          Data <- Data %>%
            mutate(Count = round(DP*EF, 0))
          #Create Factorial EF variable
          Data <- Data %>% 
            mutate(EF.Factor = EF)
          Data[Data$EF <= 0.25, ]$EF.Factor <- rep("4.LowEf", length(Data[Data$EF <= 0.25, ]$EF.Factor))
          Data[Data$EF > 0.25 & Data$EF <= 0.5, ]$EF.Factor <- rep("3.MidEf Left", length(Data[Data$EF > 0.25 & Data$EF <= 0.5, ]$EF.Factor))
          Data[Data$EF > 0.5 & Data$EF <= 0.75, ]$EF.Factor <- rep("2.MidEf Right", length(Data[Data$EF > 0.5 & Data$EF <= 0.75, ]$EF.Factor))
          Data[Data$EF > 0.75, ]$EF.Factor <- rep("1.HighEf", length(Data[Data$EF > 0.75, ]$EF.Factor))
          #Attach metadata
          Data <- Data %>%
            inner_join(., Meta, by = c("SampleID" = "Sample ID"))
          #Substitute NA for 'Control'
          Data$`KD gene ID`[is.na(Data$`KD gene ID`)] <- "Control"
          Data <- Data %>%
            as.tibble()
          if(Save == TRUE){
            write_tsv(Data, file.path(Path.Data, "vcf_comp.tsv"), col_names = TRUE)
          }
          return(Data)
        }
      }
    }
  }
}

#Load Data
Data <- Load.Data("Data/", TRUE)
```

1. Data analysis

As knockdown samples were present only for two types of cells (Feeder-free iPSC and Dermal fibroblasts) I decided to check variables inside a control group beforehand.
Firstly, I looked at Editing Fraction(EF) distribution in control group across all cell types.

```{r VP - EF, echo = FALSE}
#Plot violin plots for control
ggplot(data = Data %>%
         filter(`KD gene ID` == "Control"), 
       aes(x = `Cell description`, 
           y = EF, 
           fill = factor(`Cell description`))) +
  scale_fill_brewer(palette = "Set3") +
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) + 
  coord_flip() +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_blank(),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Editing Fraction across Cell Types", 
       fill = "Cell Type Description:")

#  geom_pointrange(stat = "summary",
#                  fun.ymin = function(z) {quantile(z, 0.25)},
#                  fun.ymax = function(z) {quantile(z, 0.75)},
#                  fun.y = median) +
#For quantiles lines
```

I observed an expectable distributions with two peaks at ~0.18 and ~0.9 to be present in all cell types.

```{r VP - EF - CvsK, echo = FALSE}
#Plot violin plots for control
ggplot(data = Data %>%
         filter(`Cell type ID` == "HDF (neonatal)" | `Cell type ID` == "FF-IPSC (IMS)"), 
       aes(x = `Cell description`, 
           y = EF, 
           fill = factor(`KD gene ID`))) +
  scale_fill_brewer(palette = "Set2") +
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) + 
  coord_flip() +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_text(angle = 45),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Editing Fraction across Cell Types", 
       fill = "Cell Type Description:")

```

The similar two-peaked pattern were present in knockdown-control pairs as well. Further, I checked distributions of sequencing depths (DP) in a similar fashion for controls and control-knockdown pairs. 

```{r Boxplots for DP, echo = FALSE}
#Boxplots for DP
ggplot(data = Data %>%
         filter(`KD gene ID` == "Control"), 
       aes(x = `Cell description`, 
           y = DP, 
           fill = factor(`Cell description`))) +
  scale_fill_brewer(palette = "Set3") +
  geom_boxplot() + 
  coord_flip(ylim = c(0, 150)) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_blank(),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Sequencing Depth across Cell Types", 
       fill = "Cell Type Description:")


```

DP medians across all samples are centered around ~25 with significant number of high-order outliers. 
Medians of knockdowns were lower when compared to paired controls, though differed not significantly from controls of other cell types.

```{r Boxplots for DP VS C, echo = FALSE}
#Boxplots for DP
ggplot(data = Data %>%
         filter(`Cell type ID` == "HDF (neonatal)" | `Cell type ID` == "FF-IPSC (IMS)"), 
       aes(x = `Cell description`, 
           y = DP, 
           fill = factor(`KD gene ID`))) +
  scale_fill_brewer(palette = "Set2") +
  geom_boxplot() + 
  coord_flip(ylim = c(0, 150)) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_text(angle = 45),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Sequencing Depth across Cell Types", 
       fill = "Cell Type Description:")

```

At the same time, the number of registered Editing Events varied considerably between cell types.

```{r Number of events, echo = FALSE}
#lollipop chart
ggplot(data = Data %>% 
         filter(`KD gene ID` == "Control") %>%
         group_by(`Cell description`) %>% 
         summarise(`Ed.Event Count` = n()), 
       aes(x = `Cell description`, 
           y = `Ed.Event Count`)) + 
  geom_point(size=3) + 
  geom_segment(aes(x =`Cell description`, 
                   xend =`Cell description`, 
                   y = 0, 
                   yend = `Ed.Event Count`)) + 
  coord_flip(ylim = c(0, 175000)) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Counts of Editing Events across Cell Types", 
       fill = "Cell Type Description:")

```

More to this, the number of editing events in knockdowns were lower in all knockdows except for XRN2 in Feeder-free iPSC.

```{r Number of Events CvsK, echo = FALSE}
#lollipop chart
library(cowplot)
plot.1 <- ggplot(data = Data %>%
         filter(`Cell type ID` == "FF-IPSC (IMS)") %>%
         group_by(`Cell description`, `KD gene ID`) %>% 
         summarise(`Ed.Event Count` = n()), 
       aes(x = `KD gene ID`, 
           y = `Ed.Event Count`)) + 
  geom_point(size=3) + 
  geom_segment(aes(x =`KD gene ID`, 
                   xend =`KD gene ID`, 
                   y = 0, 
                   yend = `Ed.Event Count`)) + 
  coord_flip(ylim = c(0, 175000)) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Ed.Events Count in Feeder-free iPSC", 
       fill = "Cell Type Description:")
plot.2 <- ggplot(data = Data %>%
         filter(`Cell type ID` == "HDF (neonatal)") %>%
         group_by(`Cell description`, `KD gene ID`) %>% 
         summarise(`Ed.Event Count` = n()), 
       aes(x = `KD gene ID`, 
           y = `Ed.Event Count`)) + 
  geom_point(size=3) + 
  geom_segment(aes(x =`KD gene ID`, 
                   xend =`KD gene ID`, 
                   y = 0, 
                   yend = `Ed.Event Count`)) + 
  coord_flip(ylim = c(0, 175000)) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.y = element_text(angle = 0),
        axis.title.y = element_blank(), 
        legend.position = "right") +
  labs(title = "Ed.Events Count in Dermal fibroblasts", 
       fill = "Cell Type Description:")
plot_grid(plot.1, plot.2, labels = "AUTO", ncol = 1, align = 'v')
```

Further, I checked the localization of observed editing events across controls of different cell types. For this purpose, I downloaded table of annotated editing sites from RADAR database and intersected it with editing data from FANTOM6 resulting in 3571 annotated sites from 990248 which I had at the beginning. 
In general, localization of editing samples was similar across all control samples. Few nonsynonymous substitutions were observed in endothelial cells, both lymphatic and blood, and in dermal fibroblasts. 

```{r PLOT ED.Count Cell-type-wise, echo = FALSE}
Data.Sites <- read_tsv("bed/Human_AG_all_hg19_v2.txt", col_names = TRUE) %>% 
  select(chromosome, position, gene, strand, annot1)
Data.Filt <- Data %>% 
  inner_join(Data.Sites, by = c("POS" = "position", "#CHROM" = "chromosome"))

#cell-type-wise barplot
ggplot(data = Data.Filt %>%
         distinct() %>%
         group_by(`Cell description`, annot1) %>%
         summarise(`Ed.Event Count` = n())) + 
  geom_bar(aes(x = `Cell description`, y = `Ed.Event Count`, fill = annot1), 
           stat = "identity") +
  scale_fill_manual(values = c("#d9d9d9","#3182bd", "#969696", "#636363", "#252525", "#e6550d")) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.x = element_text(angle = 45),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "right") +
  labs(title = "Count of Editing Events across Cell Types", 
       fill = "Cell Type Description:") +
  guides(color = FALSE)

```

```{r PLOT ED.Count Cell-type-wise (percent), echo = FALSE}
Data.Sites <- read_tsv("bed/Human_AG_all_hg19_v2.txt", col_names = TRUE) %>% 
  select(chromosome, position, gene, strand, annot1)
Data.Filt <- Data %>% 
  inner_join(Data.Sites, by = c("POS" = "position", "#CHROM" = "chromosome"))

#cell-type-wise barplot
ggplot(data = Data.Filt %>% 
         distinct() %>%
         group_by(`Cell description`, annot1) %>%
         summarise(`Ed.Event Count` = n())) + 
  geom_bar(aes(x = `Cell description`, y = `Ed.Event Count`, fill = annot1), 
           stat = "identity", position = "fill") +
  scale_fill_manual(values = c("#d9d9d9","#3182bd", "#969696", "#636363", "#252525", "#e6550d")) +
  theme_bw() + 
  theme(text = element_text(colour = "#49463C", hjust = 1, size = 14), 
        axis.text.x = element_text(angle = 45),
        axis.title.x = element_blank(), 
        axis.title.y = element_blank(),
        legend.position = "right") +
  labs(title = "Percent Editing Composition across Cell Types", 
       fill = "Cell Type Description:") +
  guides(color = FALSE)

```

Interestingly, nonsynonymous substitutions in both endothelial cells occurred in the same placce in the same gene - ENSA. This gene encodes Alpha-endosulphine protein, a member of a highly conserved cAMP-regulated phosphoprotein family.

